### PR TITLE
[Gutenberg] Add error boundary components and exception logging

### PR DESCRIPTION
**Pull request by fluiddot**
**Related PRs:**
* https://github.com/WordPress/gutenberg/pull/59221
* https://github.com/WordPress/gutenberg/pull/59385
* https://github.com/wordpress-mobile/gutenberg-mobile/pull/6655
* https://github.com/Automattic/Automattic-Tracks-Android/pull/202

This PR enables the error boundary at the editor and block level. It also connects Gutenberg Mobile to the Crash logging service in order to log JavaScript exceptions.

Additionally, it uploads to Sentry the JavaScript source map included in the React Native bridge Android library.

-----

## To Test:

Follow testing instructions from https://github.com/wordpress-mobile/gutenberg-mobile/pull/6655.

-----

## Regression Notes

1. Potential unintended areas of impact

    - None on the user-facing side as this PR mainly implements a new RN bridge function. However, it also adds a step in the build process that if fails might disrupt the release process.

2. What I did to test those areas of impact (or what existing automated tests I relied on)

    - I manually tested the build process for alpha builds. I didn't cover beta and app store builds, but the only error I foresee that it might introduce is tagging the source map files with a wrong release version. We'll work with Apps Infra team closely in the release where these changes are introduced to validate the source map files are properly uploaded.

3. What automated tests I added (or what prevented me from doing so)

    - N/A

-----

## PR Submission Checklist:

- [x] I have completed the Regression Notes.
- [ ] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

-----

## Testing Checklist:

- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] Talkback.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] Large and small screen sizes. (Tablet and smaller phones)
- [ ] Multi-tasking: Split screen and Pop-up view. (Android 10 or higher)

### CODE REVIEW

File: Gemfile
Comment (dangermattic) [inline Gemfile:12]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/f132e700785ba69cbb9320f8793e693509b3a5ba/Gemfile#L12"></span>Bundler/OrderedGems: Gems should be sorted in an alphabetical order within their section of the Gemfile. Gem <code>fastlane-plugin-sentry</code> should appear before <code>fastlane-plugin-wpmreleasetoolkit</code>.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb
Comment (dangermattic) [inline fastlane/lanes/build.rb:380]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/f132e700785ba69cbb9320f8793e693509b3a5ba/fastlane/lanes/build.rb#L380"></span>Layout/LeadingCommentSpace: Missing space after <code>#</code>.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb
Comment (dangermattic) [inline fastlane/lanes/build.rb:384]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/f132e700785ba69cbb9320f8793e693509b3a5ba/fastlane/lanes/build.rb#L384"></span>Style/QuotedSymbols: Prefer single-quoted symbols when you don't need string interpolation or special symbols.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb
Comment (dangermattic) [inline fastlane/lanes/build.rb:390]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/f132e700785ba69cbb9320f8793e693509b3a5ba/fastlane/lanes/build.rb#L390"></span>Naming/VariableName: Use snake_case for variable names.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb
Comment (dangermattic) [inline fastlane/lanes/build.rb:391]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/f132e700785ba69cbb9320f8793e693509b3a5ba/fastlane/lanes/build.rb#L391"></span>Layout/SpaceAfterComma: Space missing after comma.</td>
    </tr>
  </tbody>
</table>

Comment (dangermattic) [inline fastlane/lanes/build.rb:391]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/f132e700785ba69cbb9320f8793e693509b3a5ba/fastlane/lanes/build.rb#L391"></span>Naming/VariableName: Use snake_case for variable names.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb
Comment (dangermattic) [inline fastlane/lanes/build.rb:394]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/f132e700785ba69cbb9320f8793e693509b3a5ba/fastlane/lanes/build.rb#L394"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb
Comment (dangermattic) [inline fastlane/lanes/build.rb:405]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/f132e700785ba69cbb9320f8793e693509b3a5ba/fastlane/lanes/build.rb#L405"></span>Layout/TrailingWhitespace: Trailing whitespace detected.</td>
    </tr>
  </tbody>
</table>

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: fastlane/lanes/build.rb
Comment (fluiddot) [inline fastlane/lanes/build.rb:1]: @wordpress-mobile/apps-infrastructure I'd appreciate it if someone could take a look at this approach for uploading the React Native's JavaScript source map file to Sentry. I'm mostly seeking validation of the approach, thoughts/feedback or better alternatives. Thank you very much for the help 🙇 !

**To give some context on the approach:** The source map is generated as part of the Gutenerg Mobile build process, the files are being passed as assets via the library ([reference](https://github.com/wordpress-mobile/gutenberg-mobile/blob/a88a07389f55bcf5a233a57ff01802dbe3968a04/.buildkite/commands/publish-react-native-bridge-android-artifacts.sh#L14)). Sentry uses the JavaScript bundle and the source map to symbolicate the stack trace and reference the exact code line that caused the exception ([reference](https://docs.sentry.io/platforms/react-native/sourcemaps/uploading/hermes/)).

The approach followed here consists in:
1. Add Gem `fastlane-plugin-sentry` to use the [`sentry_upload_sourcemap` action](https://github.com/getsentry/sentry-fastlane-plugin/tree/master?tab=readme-ov-file#uploading-files--sourcemaps).
2. Add `upload_gutenberg_sourcemaps` helper function to build lanes.
3. To invoke `sentry_upload_sourcemap` we need a set of parameters that are fetched with the following steps:
    1. Load Sentry properties to obtain the token, organization and project slugs.
    2. Locate the bundle and source map files passed as assets of the [Gutenberg Mobile dependency](https://github.com/wordpress-mobile/WordPress-Android/blob/eadd3de64687891a1bbdc27a846213a8ccf97745/libs/editor/build.gradle#L86), based on app, build flavor, and build type. Note that the path points to `intermediates` folder within the project's build path.
4. Invoke the `upload_gutenberg_sourcemaps` in the different build lanes.

As a first approach, I tried to copy the bundle and source map files to a different folder during the asset merging step in the build process. However, I finally discarded this option because seemed unnecessary. [Here you can see the changes](https://github.com/wordpress-mobile/WordPress-Android/compare/rnmobile/add/log-exception-to-crash-logging...rnmobile/add/build-process-copy-bundle-source-maps).

File: fastlane/lanes/build.rb
Comment (AliSoftware) [inline fastlane/lanes/build.rb]:
```suggestion
        # When the React native bundle is generated, the source map file references include the local machine path;
        # With the `rewrite` and `strip_common_prefix` options, Sentry automatically strips this part.
```

File: fastlane/lanes/build.rb
Comment (AliSoftware) [inline fastlane/lanes/build.rb]: Any particular reason why you copy those files in a temp folder (while the files you copy seems to already have the right file names, so that's apparently not for that reason)? As opposed to directly passing `sourcemap: bundle_source_map_path` to the `sentry_upload_sourcemap` action?

Or is it because you fear there might be more sourcemap files in that `bundle_source_map_path` folder that would be accidentally uploaded if so and that's why you're isolating just those two in a temp folder?

File: fastlane/lanes/build.rb
Comment (AliSoftware) [reply fastlane/lanes/build.rb:1]: Approach looks good to me; thanks for documenting it all with comments 👍

File: fastlane/lanes/build.rb
Comment (fluiddot) [reply fastlane/lanes/build.rb]: Good observation @AliSoftware. I copied this function from iOS counterpart (https://github.com/wordpress-mobile/WordPress-iOS/pull/22655/files#diff-72b7ec258b7b79c415642e3cb618b99665e0d8e491e4b8aafe468a88636c486fR489-R493), in that case, the files need to be renamed so makes sense to copy them to a temporary folder. As you pointed out, on Android since they have the expected filenames we could directly use the original files. However, I left the copy step for the reason you mentioned about isolating the files just in case there are other sourcemap files. I'd like to note that currently there aren't others, but since the `build_asset_folder_name` contains all merged assets, this action would help us to avoid issues in the future.

File: fastlane/lanes/build.rb
Comment (fluiddot) [reply fastlane/lanes/build.rb:1]: Thanks @AliSoftware for taking a look and validating the approach, I appreciate it 🙇 .

**Review by jhnstn — APPROVED** [review]
Tested via https://github.com/wordpress-mobile/WordPress-Android/pull/20435#issuecomment-1984105872

:shipit:

**Comment by fluiddot** [comment]
I checked the `APK` generated for the installable build and noticed that the source map file is being included. Since it's not needed and the size is big (around 20MB), we should avoid include it. For this reason, I'm holding off the merge until we address this. In theory, https://github.com/wordpress-mobile/WordPress-Android/pull/20359/files#diff-276cc53d974bad3212f8e1f709d32ea3f8d380c4dccb10419505d283fb41f2d7R290 should prevent adding the file but seems it's not working.

**This can be checked by profiling the APK shared in https://github.com/wordpress-mobile/WordPress-Android/pull/20359#issuecomment-1984100770 in Android Studio**

File: fastlane/lanes/build.rb
Comment (dangermattic) [inline fastlane/lanes/build.rb]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/fe456792c3baac9e9d1e22bcc1f7c236ac54bd7e/fastlane/lanes/build.rb#L390"></span>Lint/UselessAssignment: Useless assignment to variable - <code>build_asset_folder_name</code>.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb
Comment (dangermattic) [inline fastlane/lanes/build.rb]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/fe456792c3baac9e9d1e22bcc1f7c236ac54bd7e/fastlane/lanes/build.rb#L391"></span>Layout/LeadingCommentSpace: Missing space after <code>#</code>.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/build.rb
Comment (dangermattic) [inline fastlane/lanes/build.rb]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/17ba9b95a67a3439ea8906005c572e13deb7dc64/fastlane/lanes/build.rb#L379"></span>Lint/UnusedMethodArgument: Unused method argument - <code>flavor</code>.</td>
    </tr>
  </tbody>
</table>

Comment (dangermattic) [inline fastlane/lanes/build.rb]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/17ba9b95a67a3439ea8906005c572e13deb7dc64/fastlane/lanes/build.rb#L379"></span>Lint/UnusedMethodArgument: Unused method argument - <code>build_type</code>.</td>
    </tr>
  </tbody>
</table>

File: WordPress/build.gradle
Comment (fluiddot) [inline WordPress/build.gradle:1]: Following https://github.com/wordpress-mobile/WordPress-Android/pull/20359#issuecomment-1989115970 , we had to make an adjustment in order to avoid including the source map file (`index.android.bundle.map`) in the app. This file is only needed in the build process to upload it to Sentry and shouldn't be in the resulting APK.

Originally, we tried using `exclude '**/*.bundle.map'` in this file (https://github.com/wordpress-mobile/WordPress-Android/pull/20359/commits/32db1754c193f9fb390ced5ffe01444c30b975c8) but for some reason, it was still adding the source map file. As a second approach, we are now copying the bundle and source map files to a separate folder and deleting the source map to avoid inclusion. The new folder will be used in the `upload_gutenberg_sourcemaps` function when uploading the files to Sentry.

@AliSoftware since you already took a look at the original approach (https://github.com/wordpress-mobile/WordPress-Android/pull/20359/files#r1510915719), I wonder if you could double-check that this approach is valid. Thanks for your help 🙇  !

File: fastlane/lanes/build.rb
Comment (fluiddot) [inline fastlane/lanes/build.rb:382]: This folder is created in:
https://github.com/wordpress-mobile/WordPress-Android/blob/eadc0975b7f80cc9627f031868ff73d483944cec/WordPress/build.gradle#L727-L731

File: WordPress/build.gradle
Comment (AliSoftware) [reply WordPress/build.gradle:1]: Sorry, I was focused on doing a CI deployment today, and only getting to this PR now, but I see it's already merged 🙂 

I only skimmed the new code in `build.gradle` but your explanations above makes sense to me so I think the approach sounds good 👍

Comment (fluiddot) [reply WordPress/build.gradle:1]: > Sorry, I was focused on doing a CI deployment today, and only getting to this PR now, but I see it's already merged 🙂

No worries @AliSoftware. Yes, we finally decided to merge the PR (internal ref: p1710254914686709/1710180482.367029-slack-C06G8EGHE1H) to unblock other potential PRs, as the counterpart Gutenberg changes were already merged. In any case, if we identify improvements to be made to this approach, we'll be happy to apply them in a separate PR.

> I only skimmed the new code in `build.gradle` but your explanations above makes sense to me so I think the approach sounds good 👍

Great, thanks for taking a look and validating the approach 🙇 !

File: WordPress/build.gradle
Comment (wzieba) [inline WordPress/build.gradle:738]: I have a build that fails with 
```
Could not find sourcemap at path '/var/lib/buildkite-agent/builds/ci-cluster-android-i-<id>/automattic/wordpress-android/WordPress/build/react-native-bundle-source-map'
```

https://buildkite.com/automattic/wordpress-android/builds/17693#018e51f2-20c1-48d5-80a4-ad3ccb41ef62/228-551

I think this block of code wasn't executed. @fluiddot have you seen similar issue, or maybe you have an idea why it happens?

Comment (wzieba) [reply WordPress/build.gradle:738]: I guess there was some transitive dependency bump, that's causing this problem in #20483 .

When I run `assembleJetpackJalapenoDebug` in current `trunk`, files are generated. When I do the same in #20483 , they're not.

Comment (fluiddot) [reply WordPress/build.gradle:738]: > I think this block of code wasn't executed. @fluiddot have you seen similar issue, or maybe you have an idea why it happens?
> 
> I guess there was some transitive dependency bump, that's causing this problem in #20483 .
> 
> When I run `assembleJetpackJalapenoDebug` in current `trunk`, files are generated. When I do the same in #20483 , they're not.

Hey @wzieba 👋 ! That's interesting. I haven't encountered this issue yet, and after checking other PRs, seems specific to the PR https://github.com/wordpress-mobile/WordPress-Android/pull/20483 as you pointed out.

This code block should be executed in every variant, so I wonder what's preventing it in that PR. One possibility is that the files aren't generated and so they are not copied. I can take a quick look in case I narrow down the culprit.

Comment (fluiddot) [reply WordPress/build.gradle:738]: As far as I've checked, the difference between `trunk` and https://github.com/wordpress-mobile/WordPress-Android/pull/20483 is this code block is only executed the first time. Subsequent builds don't trigger it. I only managed to trigger it again by modifying the `build.gradle` file (e.g. adding an empty line). This makes me think that there might be some kind of cache layer producing the issue. In fact, I compared the Buildkite logs between https://github.com/wordpress-mobile/WordPress-Android/pull/20483 and another one that succeeded and noticed that the task `:WordPress:mergeWordpressJalapenoDebugAssets` is marked as `FROM-CACHE` in the PR ([reference](https://buildkite.com/automattic/wordpress-android/builds/17663#018e4271-f47b-4ba1-a4a8-5a68c371292d/228-416)), whereas in other PR is not cached ([reference](https://buildkite.com/automattic/wordpress-android/builds/17691#018e51b2-f774-469f-b900-5ec1af145a8d/228-415)).

As a workaround, we could move the code block to a new Gradle task and ensure is not cached.

Comment (fluiddot) [reply WordPress/build.gradle:738]: Solving the convo as this will be fixed via https://github.com/wordpress-mobile/WordPress-Android/pull/20498.

Comment (wzieba) [reply WordPress/build.gradle:738]: Sorry @fluiddot , I only now spotted your above comments. Thanks for responding here 🙇 , and yes, we continue the discussion in #20498 .

**Pull request merged**
